### ISSUE TITLE
Handle Small/Tiny and Fast/High Speed moving object detection/tracking with stable inference

### ISSUE THREAD

**Issue by tgbaoo** [ultralytics/ultralytics#7109]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have some concern now, that I have engaged in some (fast moving - small) golf ball tracking on 2 camera frame one for swing at the tee box, one at the putting area, now I have trained the golf ball, golf hole dataset on yolov5 and yolov8 from ultralytics, but the result do not seem's too good to me because my object is the golf ball are too small.

my dataset is mostly take from this url:
https://universe.roboflow.com/anna-gaming/golfball/dataset/1

This is my training config:
```
# from ultralytics import YOLO
from IPython import display
display.clear_output()

import ultralytics
ultralytics.checks()
from ultralytics import YOLO
from IPython.display import display, Image

# model = YOLO('F:\\giabao\\golf_ball_detection_yolov8\\models\\yolov8x.pt')  # load a pretrained model (recommended for training)
# model = YOLO('yolov8s.pt')  # load a pretrained model (recommended for training)
model = YOLO('yolov8s-p2.yaml').load('/content/drive/MyDrive/Hole_In_One_Tracking/golf_ball_and_holes_detection/50k_golf_yolov8s_epoch100_batch32_imgsz720_warmup/weights/last.pt')

# model = YOLO('F:\\giabao\\golf_ball_detection_yolov8\\golf_ball_and_holes_detection\\golf_1k_epoch100_batch32_imgsz640_optAuto_gpu\\weights\\last.pt')  # load a pretrained model (recommended for training)
# model = YOLO('F:\\giabao\\golf_ball_detection_yolov8\\golf_ball_and_holes_detection\\golf_5k_epoch1_imgsz640_optAuto_gpu3\\weights\\last.pt')  # load a pretrained model (recommended for training)
# F:\\giabao\golf_ball_detection_yolov8\datasets\golf_ball_and_hole\Golf-Ball-and-Hole-Detection-6\data.yaml

results = model.train(
        project='golf_ball_and_holes_detection',
        name='50k_golf_yolov8sp2_epoch100_batch32_imgsz720',
        data='/content/drive/MyDrive/Hole_In_One_Tracking/golf_cleaned_splited_class_balanced/data.yaml',
        epochs=100,
        batch=32,
        # lr0 = 0.01,
        save=True,
        patience=200,
        # rect=True,
        imgsz=720,
        device=[0],
        # optimizer='NAdam',
        save_period	= 1,
        val=True,
        workers=12,
        resume=True,
        verbose=True,
        cache=True,
    )
```

But the last 50 epoch seem's like converging:
```
Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     50/100        28G     0.8765      0.431      1.011          1        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.88it/s]                   all       5059       5153      0.943      0.928      0.966      0.688


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     51/100      28.4G     0.8704     0.4267      1.008          2        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.87it/s]                   all       5059       5153      0.951      0.928      0.966      0.693


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     52/100      28.4G     0.8693     0.4266      1.008          1        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.88it/s]                   all       5059       5153      0.955      0.931      0.966      0.693


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     53/100      27.9G     0.8668     0.4271      1.011          2        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.87it/s]                   all       5059       5153      0.955      0.932      0.967      0.692


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     54/100        28G     0.8564     0.4234      1.003          1        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.89it/s]                   all       5059       5153      0.955      0.927      0.968      0.695


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
     55/100      27.9G     0.8532     0.4169      1.002          2        736: 100%|██████████| 1107/1107 [04:42<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 80/80 [00:27<00:00,  2.89it/s]                   all       5059       5153      0.956      0.919      0.968      0.694
```

My testing realtime data video:

https://github.com/Smorodov/Multitarget-tracker/assets/86455738/69b26320-1a24-494d-ae27-7cf178892e65

https://github.com/Smorodov/Multitarget-tracker/assets/86455738/483ef6da-e19f-4c16-9e28-59871461e716

https://github.com/Smorodov/Multitarget-tracker/assets/86455738/190ed02c-8264-42aa-912f-eb91dc279011

https://github.com/Smorodov/Multitarget-tracker/assets/86455738/4d6eb1ce-6fe6-4e63-b688-6b1e430aaef9

https://github.com/Smorodov/Multitarget-tracker/assets/86455738/b8530aa4-6dc4-42ba-93b9-a699d1fdbbfb

Is there any strategy or pipeline for the small/tiny object (high speed/fast moving) object like this?

**Comment by tgbaoo** [comment]
Plus, this is my result when applying SAHI, but It also bring a lot of false positive


https://github.com/ultralytics/ultralytics/assets/86455738/8b61a173-95a1-4205-8702-2c6fe778d47c

**Comment by tgbaoo** [comment]
yolov5 original (not free anchor ultralytics) result the weight that I use on SAHI above:
![image](https://github.com/ultralytics/ultralytics/assets/86455738/f96748c6-59f3-4f89-9ca5-a5df7994b6a6)
![image](https://github.com/ultralytics/ultralytics/assets/86455738/7181e3c5-27bd-4ed2-a8e2-c4870e36717f)
![image](https://github.com/ultralytics/ultralytics/assets/86455738/84ec22c1-ecf1-4b91-8338-5e2fa9e66d5d)
![image](https://github.com/ultralytics/ultralytics/assets/86455738/042e4710-1d18-47c4-9fe3-81ebaf3d9ddf)
![image](https://github.com/ultralytics/ultralytics/assets/86455738/006bedc9-a2bd-4492-9caa-4ffb7e2c128d)

**Comment by glenn-jocher** [comment]
@tgbaoo detecting and tracking small, fast-moving objects like golf balls can indeed be challenging. For YOLOv8, consider the following strategies to improve performance on such tasks:

1. **Annotation Quality**: Ensure your training dataset contains high-quality, accurate annotations, with tight bounding boxes around golf balls. Small objects require precise labels.

2. **Dataset Balance**: Balance your dataset with an adequate number of small object instances across various backgrounds and lighting conditions to improve generalization.

3. **Model Architecture**: Use a model architecture suited for detecting small objects. For example, the 's' models like yolov8s or yolov8n may be more suitable compared to 'x' models.

4. **Input Resolution**: Higher input resolutions can provide more details, which might be beneficial for small object detection. However, trade-offs with inference speed need consideration.

5. **Focal Loss**: Utilize focal loss which helps to focus more on hard-to-classify examples, often corresponding to small or occluded objects.

6. **Training Hyperparameters**: Experiment with hyperparameters like anchor box sizes and aspect ratios, learning rate, and augmentation settings, to suit the specifics of your small object detection task.

7. **Data Augmentation**: Apply appropriate augmentations like mixup, cutmix, or mosiac, which might force the model to focus on smaller regions and could potentially improve small object detection.

8. **Post-processing**: Use NMS thresholds and confidence scores calibrated to filter out false positives without removing true positives.

9. **Tracking**: For accurate tracking, consider integrating YOLOv8 with dedicated tracking algorithms optimized for small and fast-moving objects.

Addressing false positives with SAHI or post-processing requires fine-tuning the confidence thresholds and possibly adding custom filtering logic based on object size or predicted motion vectors.

For further insights and troubleshooting, kindly refer to the [Ultralytics Docs](https://docs.ultralytics.com), which provide comprehensive guidance on model training and optimization.

**Comment by tgbaoo** [comment]
@glenn-jocher Thanks for your reply, could you give me some strategy for 6, 7 and 8? If you can, from your experience, could you share some tracking algorithm you think could fit this problem at 9?

**Comment by tgbaoo** [comment]
@glenn-jocher I saw you in a lot of post that modify the anchor box size, but I do not see where contain that, and other param above, could you give me some detail place where to find? Thank you so much, forgive me as a newbie on computer vision

**Comment by glenn-jocher** [comment]
@tgbaoo certainly, for strategies 6, 7, and 8:

6. **Training Hyperparameters**: Adjust the `anchor` settings in the model configuration YAML file to better suit the scale of small objects in your dataset. Also, experiment with the learning rate schedule to find the optimal setting where the model learns fine features without overfitting.

7. **Data Augmentation**: In your training configuration settings, you can alter augmentation parameters such as `mosaic`, `mixup`, `hsv_h`, `hsv_s`, `hsv_v`, `degrees`, and `translate`. Carefully tuned augmentations can help the model generalize better to small objects.

8. **Post-processing**: Modify the Non-Maximum Suppression (NMS) parameters such as `iou_thresh` and `conf_thresh` to filter detections. This can be done during validation/training or in your inference pipeline.

For tracking (strategy 9):

- **SORT (Simple Online and Realtime Tracking)** is a good starting point because of its simplicity and efficiency.
- **DeepSORT** extends SORT with deep learning features, providing improved tracking capabilities, which can be quite effective for tracking smaller objects when combined with a robust detection model like YOLOv8.

To modify anchor box sizes and other hyperparameters, you can locate the configuration file (`*.yaml`) used for your model. It contains various settings including anchor box definitions, which you can adjust based on your object sizes and aspect ratios. In the YOLOv8 repo, this file is typically named following the format `yolov8*.yaml` and can be found in the `models` directory.

For a more in-depth explanation and best practices, please refer to the documentation on [Train](https://docs.ultralytics.com/modes/train) and [Val](https://docs.ultralytics.com/modes/val) modes, which covers hyperparameter tuning and anchor box adjustments as well. Please do not hesitate to reach out if you have further questions.

**Comment by tgbaoo** [comment]
@glenn-jocher Sorry but, I have investigate the directory ```ultralytics/cfg/models/v8/yolov8.yaml```
but i just not found the anchor box param on this yaml file and other in the same directory:

yolov8.yaml:
```
# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv8 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80  # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]  # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024]  # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768]   # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512]   # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512]   # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

  - [[15, 18, 21], 1, Detect, [nc]]  # Detect(P3, P4, P5)
```

**Comment by glenn-jocher** [comment]
@tgbaoo, in YOLOv8, anchors are dynamically calculated during training and are not manually set in the model configuration file as in previous versions. The model determines the best anchor boxes to fit your specific dataset during the training process. For more information on the dynamic anchor calculation and other YOLOv8 features, kindly refer to the comprehensive [Ultralytics Docs](https://docs.ultralytics.com). If you have further questions or require additional guidance, feel free to reach out. Happy training! 🚀

**Comment by tgbaoo** [comment]
Hello @glenn-jocher, now the detection step seem's good. But I am having another issue is about the object tracking task especially for golf ball object(not human label - that trained on ReID model for tracking algorithms). I have research but do not have any documentation instruct me for training a ReID model for custom object.
I am planning to train a ReID model for my custom Golf Ball dataset to enhanced the ability to tracking and re-identification the golf ball object. 
If you have any idea or from your experience, could you share with my your opinion about my idea? Is that possible? If not, after training yolov8, which step or algorithm should I take notice to finish my object tracking task?

My detection result with StrongSORT:
https://github.com/ultralytics/ultralytics/assets/86455738/9a39c7af-cbc6-4609-9e9c-877bbbdffbab

this is my training result:
![train_batch99630](https://github.com/ultralytics/ultralytics/assets/86455738/434d6324-af9f-4a9e-97c8-355bafd0c50f)
![train_batch99631](https://github.com/ultralytics/ultralytics/assets/86455738/376e4695-8f34-4c7e-b2e7-3d5f434ef120)
![train_batch99632](https://github.com/ultralytics/ultralytics/assets/86455738/8b8fd182-be08-4e97-88b4-8bd9234354d7)
![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/86455738/a659b0fe-b320-460e-ba48-26044ad4c276)
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/86455738/458275de-8431-4726-a0bd-3d596b4426fe)
![val_batch1_labels](https://github.com/ultralytics/ultralytics/assets/86455738/36b6e5f2-d6f4-4ed6-ac84-6b1fd3fa3277)
![val_batch1_pred](https://github.com/ultralytics/ultralytics/assets/86455738/a21823e2-f6d9-4d97-9bbe-923469c40e27)
![val_batch2_labels](https://github.com/ultralytics/ultralytics/assets/86455738/322b46df-23b4-4ca0-99f4-8fcb4ab8e294)
![val_batch2_pred](https://github.com/ultralytics/ultralytics/assets/86455738/7c217699-ae91-46fd-97d8-83b7a4485434)
![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/86455738/3c7dbf81-992d-4fea-80db-b17c5cb93c60)
![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/86455738/6bb34bef-d629-4189-8ebc-b84535ce24e5)
![labels](https://github.com/ultralytics/ultralytics/assets/86455738/e4c709bf-7138-4be2-9118-e55e57b532e6)
![train_batch1](https://github.com/ultralytics/ultralytics/assets/86455738/9e790ceb-5e21-41a0-9ea5-1644d86194c8)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/86455738/58a9aa18-2a55-4766-a2c3-aaa23682d132)
![labels_correlogram](https://github.com/ultralytics/ultralytics/assets/86455738/c67cdfe6-9f00-4107-a545-c8fc862b48a3)
![P_curve](https://github.com/ultralytics/ultralytics/assets/86455738/02d525fc-81c9-4447-aa7a-3a0ec787d945)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/86455738/9e470ce4-68f2-42d2-b013-3932150e3089)
![R_curve](https://github.com/ultralytics/ultralytics/assets/86455738/7a2592a7-aa42-411f-9333-66a0edee4c38)
![results](https://github.com/ultralytics/ultralytics/assets/86455738/4098c256-31d4-4263-b979-6c1415884d09)
![train_batch0](https://github.com/ultralytics/ultralytics/assets/86455738/bd047460-c890-48ce-9c95-f05c2211e2d1)
![train_batch2](https://github.com/ultralytics/ultralytics/assets/86455738/4d10cbb2-bfaf-4d1a-8ccc-d4da78433ebe)

**Comment by glenn-jocher** [comment]
@tgbaoo, training a ReID model for custom objects like golf balls is indeed possible and can enhance tracking performance. However, it requires a dataset with multiple instances of the same object across different scenes or conditions to learn distinctive features.

For your golf ball tracking task, after training YOLOv8, you might consider the following steps:
- **Data Collection**: Gather a diverse set of images with golf balls under various conditions and from different angles.
- **Feature Extraction**: Use a pre-trained model or train a new one to extract features from detected golf balls.
- **Metric Learning**: Implement a metric learning approach, such as triplet loss, to train your ReID model to differentiate between different golf balls.
- **Integration**: Combine your ReID model with a tracking algorithm like StrongSORT, which you're already using, to maintain identities over time.

If creating a ReID model is not feasible, focus on fine-tuning your detection model and tracking algorithm parameters to reduce identity switches and improve tracking consistency. For more detailed guidance, the [Ultralytics Docs](https://docs.ultralytics.com) provide valuable resources on model training and deployment. Keep up the good work, and don't hesitate to ask if you need further assistance! 🌟

**Comment by tgbaoo** [comment]
@glenn-jocher Like I do not have any information about how to train or how to prepare a dataset structure about training custom object on ReID model, so if you joined any relevant project please give me some keywords or some documentation on this field?

Because I have tried my best on detection model and fine-tuning the SORT algorithm's parameters, so ReID is my last choice.

**Comment by glenn-jocher** [comment]
@tgbaoo, for training a custom ReID model, focus on collecting a dataset where each individual object (golf ball) appears multiple times across different scenes. Label each instance with a unique ID across all scenes. Use a CNN to extract features and employ a metric learning approach, like triplet loss, to distinguish between different objects. Look into papers and resources on "deep metric learning," "triplet loss for ReID," and "person re-identification" for methodologies that can be adapted to objects. Unfortunately, detailed documentation for ReID on custom non-human objects is limited, so you may need to adapt strategies from related fields. Good luck with your project! 🚀

**Comment by tgbaoo** [comment]
Thanks you for the comments, I have found this repo and implementing this repo with yolov8 to working on my project:
https://github.com/mikel-brostrom/yolo_tracking/issues/1246

**Comment by glenn-jocher** [comment]
@tgbaoo, you're welcome! I'm glad to hear that you've made progress by integrating YOLOv8 with the tracking repository. A big shout-out to @mikel-brostrom for the support! Keep up the great work on your project, and if you have any more questions or need further assistance, feel free to reach out. Good luck with the next stages! 🎉🚀

**Comment by AInsight-AI** [comment]
> @tgbaoo detecting and tracking small, fast-moving objects like golf balls can indeed be challenging. For YOLOv8, consider the following strategies to improve performance on such tasks:
> 
> 1. **Annotation Quality**: Ensure your training dataset contains high-quality, accurate annotations, with tight bounding boxes around golf balls. Small objects require precise labels.
> 2. **Dataset Balance**: Balance your dataset with an adequate number of small object instances across various backgrounds and lighting conditions to improve generalization.
> 3. **Model Architecture**: Use a model architecture suited for detecting small objects. For example, the 's' models like yolov8s or yolov8n may be more suitable compared to 'x' models.
> 4. **Input Resolution**: Higher input resolutions can provide more details, which might be beneficial for small object detection. However, trade-offs with inference speed need consideration.
> 5. **Focal Loss**: Utilize focal loss which helps to focus more on hard-to-classify examples, often corresponding to small or occluded objects.
> 6. **Training Hyperparameters**: Experiment with hyperparameters like anchor box sizes and aspect ratios, learning rate, and augmentation settings, to suit the specifics of your small object detection task.
> 7. **Data Augmentation**: Apply appropriate augmentations like mixup, cutmix, or mosiac, which might force the model to focus on smaller regions and could potentially improve small object detection.
> 8. **Post-processing**: Use NMS thresholds and confidence scores calibrated to filter out false positives without removing true positives.
> 9. **Tracking**: For accurate tracking, consider integrating YOLOv8 with dedicated tracking algorithms optimized for small and fast-moving objects.
> 
> Addressing false positives with SAHI or post-processing requires fine-tuning the confidence thresholds and possibly adding custom filtering logic based on object size or predicted motion vectors.
> 
> For further insights and troubleshooting, kindly refer to the [Ultralytics Docs](https://docs.ultralytics.com), which provide comprehensive guidance on model training and optimization.

Hey Glenn, is the YOLOs means smaller model for yolo? or that means detecting for small object? I am not sure can you please double confirm?

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #19870, which introduces the new CutMix image augmentation feature in Ultralytics 8.3.119. With CutMix, users can now apply an advanced augmentation technique that improves model robustness by simulating realistic occlusions and challenging scenarios during training. This update adds a cutmix option alongside existing methods like mosaic and mixup, updates configuration files, and enhances the documentation with detailed guides and visuals. The integration is seamless, allowing flexible tuning and easy adoption within the training pipeline.

Special thanks to the PR author and contributors for their efforts in bringing this feature to Ultralytics.

To test if this resolves your issue, please update your package with:
```
pip install -U ultralytics>=8.3.119
```
Alternatively, you can install the latest code from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if you prefer.

Your feedback is appreciated—please let us know if this update resolves the problem or if you encounter any further issues. Thank you for reporting this and helping improve Ultralytics! 🙏
